src/appAccess.ts

```diff
@@ -0,0 +1,288 @@
+      if (
+        a.get('d') === target.address &&
+        a.get('s') === target.assetIndex &&
+        a.get('p') === target.appIndex
+      ) {
+        return idx + 1; // 1-based index
+      }
+    }
+    if (target.address) {
+      accessList.push(new Map<string, unknown>([['d', target.address]]));
+    }
+    if (target.assetIndex) {
+      accessList.push(new Map<string, unknown>([['s', target.assetIndex]]));
+    }
+    if (target.appIndex) {
+      accessList.push(new Map<string, unknown>([['p', target.appIndex]]));
+    }
+    return accessList.length; // length is 1-based position of new element
+  }
+
+  const zeroAddr = Address.zeroAddress();
+  for (const rr of references) {
+    if (rr.address || rr.assetIndex || rr.appIndex) {
+      ensure(rr);
+    }
+
+    if (rr.holding) {
+      const h = rr.holding;
+      let addrIdx = 0;
+      if (h.address && !(h.address as Address).equals(zeroAddr)) {
+        addrIdx = ensure({ address: h.address });
+      }
+      const assetIdx = ensure({ assetIndex: h.assetIndex });
+      accessList.push(
+        new Map<string, unknown>([
+          [
+            'h',
+            new Map<string, unknown>([
+              ['d', addrIdx],
+              ['s', assetIdx],
+            ]),
+          ],
+        ])
+      );
+    }
+    if (rr.locals) {
+      const l = rr.locals;
+      let addrIdx = 0;
+      if (l.address && !(l.address as Address).equals(zeroAddr)) {
+        addrIdx = ensure({ address: l.address });
+      }
+      let appIdx = 0;
+      if (l.appIndex && BigInt(l.appIndex) !== appIndex) {
+        appIdx = ensure({ appIndex: l.appIndex });
+      }
+      accessList.push(
+        new Map<string, unknown>([
+          [
+            'l',
+            new Map<string, unknown>([
+              ['d', addrIdx],
+              ['p', appIdx],
+            ]),
+          ],
+        ])
+      );
+    }
+    if (rr.box) {
+      const b = rr.box;
+      let appIdx = 0;
+      if (b.appIndex && BigInt(b.appIndex) !== appIndex) {
+        appIdx = ensure({ appIndex: b.appIndex });
+      }
+      accessList.push(
+        new Map<string, unknown>([
+          [
+            'b',
+            new Map<string, unknown>([
+              ['i', appIdx],
+              ['n', b.name],
+            ]),
+          ],
+        ])
+      );
+    }
+  }
+  return accessList;
+}
+
+export function convertIndicesToResourceReferences(
+  accessList: Array<Map<string, unknown>>
+): Array<ResourceReference> {
+  const references: Array<ResourceReference> = [];
+  for (const item of accessList) {
+    const address = item.get('d') as Address | undefined;
+    const assetIndex = item.get('s') as bigint | undefined;
+    const appIndex = item.get('p') as bigint | undefined;
+    if (address) {
+      references.push({ address });
+    } else if (assetIndex) {
+      references.push({ assetIndex });
+    } else if (appIndex) {
+      references.push({ appIndex });
+    }
```

Comment:
should we `continue` these pushes, so the reader knows there's nothing else going on?

Reply:
done